clio-server/src/main/scala/org/broadinstitute/clio/server/service/IndexService.scala

```diff
@@ -67,4 +78,133 @@ abstract class IndexService[CI <: ClioIndex](
       elasticsearchIndex
     )
   }
+  private val keysToDrop =
+    Set(
+      ElasticsearchIndex.UpsertIdElasticsearchName,
+      ElasticsearchIndex.EntityIdElasticsearchName,
+    )
+
+  def validateUpsert(
+    indexKey: clioIndex.KeyType,
+    metadata: clioIndex.MetadataType,
+    force: Boolean
+  )(
+    implicit materializer: Materializer
+  ): Either[Exception, clioIndex.MetadataType] = {
+
+    // drop housekeeping and key fields (evaluate only metadata)
+    val dropKeys = keysToDrop ++ indexKey.getClass.getDeclaredFields
+      .map(field => ElasticsearchUtil.toElasticsearchName(field.getName))
+
+    val storedDocsForKey = Await
+      .result(
+        searchService
+          .simpleStringQueryMetadata(indexKey.asJson.pretty(Printer.noSpaces))(
+            elasticsearchIndex
+          )
+          .runWith(Sink.seq),
```

Comment:
`Sink.seq` should be safe in practice, but if we majorly mess up it might collect a huge list in-memory and crash the server. In my client-side streaming branch I used a [stream fold](https://github.com/broadinstitute/clio/blob/dm-client-streaming-DSDEGP-1985/clio-client/src/main/scala/org/broadinstitute/clio/client/webclient/ClioWebClient.scala#L160) to avoid building up the huge list.

Or, we could lean on the key invariant and use `Sink.headOption`.